### ISSUE TITLE
`T | None` does not work when T is an union

### ISSUE THREAD

**Issue by Hakuyume** [python/mypy#20243]
**Bug Report**

mypy can not handle `T | None` when `T` is also an union.

**To Reproduce**

Run mypy on the following code.

```python
def unwrap[T](value: T | None) -> T:
    assert value is not None
    return value


def f(value: int | str) -> None:
    pass


def main() -> None:
    value: int | str | None = 42
    f(unwrap(value))
```
https://mypy-play.net/?mypy=latest&python=3.12&gist=7c62868a7c9a0afc4853487fdd605f4b

**Expected Behavior**

mypy does not show any errors. 
`unwrap(value)` converts `int | str | None` to `int | str`.

**Actual Behavior**

```
main.py:12: error: Argument 1 to "f" has incompatible type "object"; expected "int | str"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

mypy handles `unwrap(value)` as `object` and raises an error.

**Your Environment**

- Mypy version used: 1.18.2
- Mypy command-line flags: default of playground
- Mypy configuration options from `mypy.ini` (and other config files): default of playground
- Python version used: 3.12

**Comment by A5rocks** [comment]
This is yet another case where mypy using join is confusing (compared to using a union). Probably it would make sense to switch this? (I think this is the case?)

As a workaround, you can rely on type context. I think this should work?
```
x: int | str = unwrap(value)
```

(If it doesn't then that's a bug IMO and should get its own issue)

---

Actually maybe mypy already does use type context for the original code snippet (I can't remember the rules...). Something that definitely doesn't is:

```
x = unwrap(value)
f(x)  # should not error
```

**Comment by Hakuyume** [comment]
@A5rocks Thank you for sharing workaround. I tried the following code but it does not work.

```python
def unwrap[T](value: T | None) -> T:
    assert value is not None
    return value


def f(value: int | str) -> None:
    pass


def main() -> None:
    value: int | str | None = 42
    x: int | str = unwrap(value)
    f(x)
```
https://mypy-play.net/?mypy=latest&python=3.12&gist=26813015e91e6cfb3fc0e1fbd2452e4c

```
main.py:12: error: Incompatible types in assignment (expression has type "object", variable has type "int | str")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

**Comment by ilevkivskyi** [comment]
I think this is a duplicate of https://github.com/python/mypy/issues/17383, let's continue discussion there.
